crates/ark/src/lsp/completions/builder.rs

```diff
@@ -0,0 +1,73 @@
+//
+// builder.rs
+//
+// Copyright (C) 2025 Posit Software, PBC. All rights reserved.
+//
+//
+
+use std::cell::OnceCell;
+
+use anyhow::Result;
+use tower_lsp::lsp_types::CompletionItem;
+
+use crate::lsp::completions::parameter_hints::ParameterHints;
+use crate::lsp::completions::parameter_hints::{self};
+use crate::lsp::completions::sources::composite::find_pipe_root;
+use crate::lsp::completions::sources::composite::pipe::PipeRoot;
+use crate::lsp::completions::sources::composite::CompositeCompletionsSource;
+use crate::lsp::completions::sources::unique::UniqueCompletionsSource;
+use crate::lsp::completions::sources::CompletionSource;
+use crate::lsp::document_context::DocumentContext;
+use crate::lsp::state::WorldState;
+
+pub(crate) struct CompletionBuilder<'a> {
+    pub(crate) context: &'a DocumentContext<'a>,
+    pub(crate) state: &'a WorldState,
+    parameter_hints_cell: OnceCell<ParameterHints>,
+    pipe_root_cell: OnceCell<Option<PipeRoot>>,
+}
+
+impl<'a> CompletionBuilder<'a> {
+    pub fn new(context: &'a DocumentContext, state: &'a WorldState) -> Self {
+        Self {
+            context,
+            state,
+            parameter_hints_cell: OnceCell::new(),
+            pipe_root_cell: OnceCell::new(),
+        }
+    }
+
+    pub fn parameter_hints(&self) -> &ParameterHints {
+        self.parameter_hints_cell.get_or_init(|| {
+            parameter_hints::parameter_hints(self.context.node, &self.context.document.contents)
+        })
+    }
+
+    pub fn pipe_root(&self) -> Result<Option<PipeRoot>> {
+        if let Some(root) = self.pipe_root_cell.get() {
+            return Ok(root.clone());
+        }
+
+        let root = find_pipe_root(self.context)?;
+
+        // Cache it for future calls (ignore failure if race condition, which shouldn't happen)
+        let _ = self.pipe_root_cell.set(root.clone());
+
+        Ok(root)
+    }
+
+    pub fn build(self) -> Result<Vec<CompletionItem>> {
```

Comment:
It feels to me like something is wrong here. `build()` calls into the aggregators who then call into the sources with the builder passed in. So a source could potentially call `build()` again, which doesn't feel right?

I think we're incorrectly forcing the builder pattern here? It's more like we only have a context that is passed around. So I'd rename to "state" or "context" and make that `build()` method a free function that is not accessible from sources?

Reply:
Done.